payjoin/src/receive/v2.rs

```diff
@@ -209,33 +209,9 @@ impl Enrolled {
             log::debug!("response is empty");
             return Ok(None);
         }
-        // parse v1 or v2 proposal
         match String::from_utf8(response.clone()) {
-            Ok(proposal) => {
-                let context = V2Context {
-                    directory: self.directory.clone(),
-                    ohttp_keys: self.ohttp_keys.clone(),
-                    ohttp_relay: self.ohttp_relay.clone(),
-                    s: self.s,
-                    e: None,
-                };
-                log::debug!("Received proposal: {}", proposal);
-                Ok(Some(UncheckedProposal::from_v2_payload(proposal.into_bytes(), context)?))
-            }
-            Err(_) => {
-                let (proposal, e) = crate::v2::decrypt_message_a(&response, self.s.secret_key())?;
-                log::debug!("Some e: {}", e);
-                let context = V2Context {
-                    directory: self.directory.clone(),
-                    ohttp_keys: self.ohttp_keys.clone(),
-                    ohttp_relay: self.ohttp_relay.clone(),
-                    s: self.s,
-                    e: Some(e),
-                };
-                let proposal = UncheckedProposal::from_v2_payload(proposal, context)?;
-
-                Ok(Some(proposal))
-            }
+            Ok(response) => Ok(Some(self.extract_proposal_from_v1(response)?)),
+            Err(_) => Ok(Some(self.extract_proposal_from_v2(response)?)),
```

Comment:
If I understand correctly, this is expected to be `Ok` if the response is plaintext and `Err` if it's encrypted?

This may be more readable:

```
match crate::v2::decrypt_message_a(&response, self.s.secret_key()) {
    // v2 payload
    Ok(payload_bytes, e) => Ok(Some(self.extract_proposal(payload_bytes, e)?)),
    // v1 payload
    Err(_) => Ok(Some(self.extract_proposal(response, None)?)),
}

fn extract_proposal(&self, response: Vec<u8>, e: Option<PublicKey>) -> Result<UncheckedProposal, Error> {
    let context = self.build_context(e);
    let payload = String::from_utf8(response).map_err(InternalRequestError::Utf8)?;
    Ok(UncheckedProposal::from_payload(payload, context)?)
}
```

Reply:
I took a stab [implementing your suggestion](https://github.com/payjoin/rust-payjoin/commit/3a32bdfed6cc1d8c9e5a30a108fd2b9fac74100c) but I think it loses correctness since it throws out all decryption errors and tries to decode as a v1 body even if it could be a crypto error and not a body mismatch. 

When decryption fails with this suggestion, any `decrypt_message_a` error is thrown out. The implicit assumption is that the payload must be v1 because decryption failed, when it could fail for a number of reasons.

However in 35e38dbd88b9196b61bc3a135fcd7ecfecb27eeb, matching over String decoding, the only possibility is that the body is not a string and thus cannot be a valid v1 response, so v2 is tried. Then v2 is processed and any other v2-related decryption error may be propagated completely using `?`.

```
        match String::from_utf8(response.clone()) {
            // V1 response bodies are utf8 plaintext
            Ok(response) => Ok(Some(self.extract_proposal_from_v1(response)?)),
            // V2 response bodies are encrypted binary blobs
            Err(_) => Ok(Some(self.extract_proposal_from_v2(response)?)),
        }
```

I added the above comments to explain. Does this satisfy your review? Any other suggestion for legibility?

Reply:
That makes sense, thanks for the detailed response